File: ParallelAndVectorized/Algorithms/OptimizedVectorizedMatrixMultiplication.py
import numpy as np
from numba import jit

class OptimizedVectorizedMatrixMultiplication:
    @staticmethod
    @jit(nopython=True)
    def _vectorized_multiply(matrixA, matrixB):
        rows_A, cols_A = matrixA.shape
        rows_B, cols_B = matrixB.shape

        if cols_A != rows_B:
            raise ValueError("The number of columns in matrix A must be equal to the number of rows in matrix B.")

        result = np.zeros((rows_A, cols_B), dtype=np.float32)

        for i in range(rows_A):
            for j in range(cols_B):
                result[i, j] = np.dot(matrixA[i, :], matrixB[:, j])

        return result

    def multiply(self, matrixA, matrixB):
        return self._vectorized_multiply(matrixA, matrixB)

File: ParalellAndVectorizedMatrices/src/main/java/org/example/OptimizedVectorizedMatrixMultiplication.java
package org.example;

public class OptimizedVectorizedMatrixMultiplication implements MatrixMultiplication{

    @Override
    public double[][] multiply(double[][] matrixA, double[][] matrixB) {
        int rowsA = matrixA.length;
        int colsA = matrixA[0].length;
        int rowsB = matrixB.length;
        int colsB = matrixB[0].length;

        if (colsA != rowsB) {
            throw new IllegalArgumentException("The number of columns in matrix A must be equal to the number of rows in matrix B.");
        }

        double[][] result = new double[rowsA][colsB];

        for (int i = 0; i < rowsA; i++) {
            for (int j = 0; j < colsB; j++) {
                result[i][j] = vectorizedDotProduct(matrixA[i], getColumn(matrixB, j));
            }
        }

        return result;
    }

    private static double vectorizedDotProduct(double[] row, double[] column) {
        double sum = 0;
        for (int k = 0; k < row.length; k++) {
            sum += row[k] * column[k];
        }
        return sum;
    }

    private static double[] getColumn(double[][] matrix, int colIndex) {
        double[] column = new double[matrix.length];
        for (int i = 0; i < matrix.length; i++) {
            column[i] = matrix[i][colIndex];
        }
        return column;
    }
}
